### PR TITLE
Show proposal details

### PR DESCRIPTION
**Pull request by jasonaowen**
Add a new proposal detail page at `/proposal/:proposalId` that will fetch and display the details of the most recent version of that proposal. Visiting that page will automatically start the log in process if you are not yet logged in.

This is the first example of a pattern I'm proposing of separating display components (such as `<ProposalDetailPanel>`) from logic components that make API requests, transform API responses into the format display components expect, and so on. This separation will, I hope, make it easier to prototype new components, play with components in the storybook, and disentangle display logic from business logic. (This PR does not introduce any display logic, but an example would be the button that toggles between showing canonical field names and short codes; I'd expect the display components to accept all the data they'd need, and then internally decide what to display.)

Note that to use this locally, you will need to define the new `REACT_APP_API_URL` environment variable.

This draws heavily from PR #44, but leaves some things behind (such as buttons that are not yet implemented) and adds functionality.

### CODE REVIEW

File: src/pages/ProposalDetailPage.tsx
Comment (jasonaowen) [inline src/pages/ProposalDetailPage.tsx:85]: @jim-mcgowan I'd appreciate your thoughts on this priority list for picking canonical fields to fill the header! In case the code is unclear, it picks the first present value, so for title it'd look for `proposal_name`, then if that's absent `proposal_title`, then if that's absent it'll construct a name like `Proposal 7`.

**Comment by jasonaowen** [comment]
I added the environment variable to Netlify, and added the deploy preview to the Keycloak client configuration.

A few example pages:
- [Proposal 5](https://deploy-preview-47--philanthropy-data-commons-viewer.netlify.app/proposal/5) has all the information we're looking for
- [Proposal 7](https://deploy-preview-47--philanthropy-data-commons-viewer.netlify.app/proposal/7) has no title
- [Proposal 0](https://deploy-preview-47--philanthropy-data-commons-viewer.netlify.app/proposal/0) does not exist, and will stay on the loading screen forever, as mentioned in the PR description
- The [Proposal Detail Storybook](https://deploy-preview-47--philanthropy-data-commons-viewer.netlify.app/storybook/?path=/story/stories-proposaldetail--default) does not require logging in, and allows you to play with the component

File: src/components/ProposalTable.tsx
Comment (reefdog) [inline src/components/ProposalTable.tsx:38]: Ah, is `position` actually unique in each dataset (e.g. safe to use for `key` on its own rather than as a string like `${fieldName}-${position}`)? I assumed it was relative to other fields with the same fieldName. I assume you're right, but just flagging it in case.

File: src/pages/ProposalDetailPage.tsx
Comment (reefdog) [inline src/pages/ProposalDetailPage.tsx:96]: I'm a little nervous at how quickly we're going to want to replace this architecture when we really support versions, but you noted that limitation in the description and I think it's the right choice. Interested to know if you already have the adaptations in mind!

File: src/use-pdc-api.ts
Comment (reefdog) [inline src/use-pdc-api.ts:31]: I _think_ this implementation misuses effect dependencies.

This is treating them like function parameters, but the `useEffect` hook has access to the outer scope and doesn't need that (which is why `fetch` works). These are dependencies that should trigger the effect to run again when they change (which is why including `fetch` in the dependencies causes it to repeatedly re-run).

Assuming we only want this to run on initial render, we still want the dependency array, but I think it should be empty, right?

```suggestion
  }, []);
```

I may be totally misunderstanding what's going on here, but I don't quite follow as-is.

**Review by reefdog — CHANGES_REQUESTED** [review]
This is wonderful work! I'm not currently able to run it locally because I just realized I don't know how to authenticate into the PDC from localhost (or if that's even possible), but looking at the deployment preview it's a great start.

I do have a scattering of individual comments and requests. If you resolve them to your satisfaction, you can feel free to merge, or I can take a second look on Monday.

**Comment by jim-mcgowan** [comment]
@jasonaowen I need to see this to be able to comment.

File: .env.example
Comment (jasonaowen) [reply .env.example:24]: I did, but I'm very open to the feedback that I should stick to the existing pattern - I will fix!

File: src/pages/ProposalDetailPage.tsx
Comment (jasonaowen) [reply src/pages/ProposalDetailPage.tsx:85]: I did consider writing a helper function, like "get value of first present field" or so, but it's just a few lines of code and this seems clear enough for now.

...but I bet we could do it with a map, a filter, and maybe a reduce!

File: src/pages/ProposalDetailPage.tsx
Comment (jasonaowen) [reply src/pages/ProposalDetailPage.tsx:96]: I do! I'm sure what I have in mind will change a bit as I implement it, but I think I have a good enough plan. And yeah these'll change pretty quick, but I think that's okay!

File: src/components/ProposalDetailPanel.tsx
Comment (jasonaowen) [reply src/components/ProposalDetailPanel.tsx:17]: That sounds like a better name, thank you! I will fix.

File: src/use-pdc-api.ts
Comment (jasonaowen) [reply src/use-pdc-api.ts:31]: Including `setResponse` is probably being over-cautious, yes. But I think we do need to include `path`, because otherwise we can't navigate without a full page reload.

I tried it by adding prev/next buttons, and removing `path` from the dependency array, and navigating to another proposal did not load any new data. I believe this is because we have the same React component tree, and the only difference React sees is the new react-router path parameter ([via the `useParams()` hook](https://reactrouter.com/en/main/hooks/use-params)). We use that to construct the API relative path, but if we do not include that relative path (the `path` function parameter here) in the dependency list, React looks and says "I don't need to re-run this effect", and so we do not update the data.

I *think* we would ideally want `fetch` to be in the dependency list, too, but I'm not completely certain. I would expect it to trigger the effect to run once its token expires and is refreshed, but that's not how it works currently (see the upstream bu g I filed), and even so that's kinda getting into IdP configuration territory - if we set very short expiration times on our tokens, do we want viewers to continuously be making API calls? Probably not, so maybe the comment is incomplete...

File: src/pages/ProposalDetailPage.tsx
Comment (jasonaowen) [reply src/pages/ProposalDetailPage.tsx:85]: I talked with @jim-mcgowan on the phone, and my takeaway was that this is good for now, so resolving.

File: src/components/ProposalDetailPanel.tsx
Comment (jasonaowen) [inline src/components/ProposalDetailPanel.tsx:45]: After talking with @jim-mcgowan, his main feedback was that we should be highlighting the applicant above the proposal title. The EIN is of secondary importance, and should not be given the same weight as the promoted organization name; he suggested displaying it below, but maybe alongside in gray might work too - I'll play with it a bit, or possibly wait for @reefdog's feedback.

File: src/components/ProposalDetailPanel.tsx
Comment (jasonaowen) [inline src/components/ProposalDetailPanel.tsx:41]: @jim-mcgowan was not a fan of this users icon: "I'm not sure what that's trying to convey. Three people would suggest an organization to me."

There is a `user-group` in heroicons; I'll try that.

**Review by jasonaowen — COMMENTED** [review]
@jim-mcgowan and I spoke on the phone, and he shared some feedback. I believe I've documented all the front end concerns below; there were also some data concerns that I'll document elsewhere.

File: src/components/ProposalTable.tsx
Comment (jasonaowen) [reply src/components/ProposalTable.tsx:38]: Excellent question! I was slightly confident when I committed, but after doing some more reading and thinking, I am now more confident. (One of the better articles I've found was [React key attribute: best practices for performant lists](https://www.developerway.com/posts/react-key-attribute).)

I think the two key (hah) things to think about here is what state items in the list carry, when this list will change. 

Maintaining state across reordering or complete switches of underlying data is important to maintain correct behavior. The example from the article is instructive: if you have the first item in the list checked, and then change the sort, the checkmark should follow the item rather than the index. However, in our case, list items (`<TableRow>`s, here) do not carry any state, so I do not think there is a risk of incorrect behavior no matter what approach we take.

If correctness is not a concern, then that leaves performance. I believe this list will change under two circumstances:  when navigating from one proposal to another or toggling the version history. The version history is simpler: turning it on can only add rows at the end (fields that existed on previous versions of the proposal but not the current version of the proposal), and turning it off again will only delete rows at the end, so we could even use the `index` and be okay.

Switching proposals is more complicated, but reusing positions will allow React to maintain most of the existing DOM structure, changing only the values. Some positions might be missing, and others added (eg a proposal for an application form with field positions `[1, 2, 4]` switching to one with positions `[1, 2, 3, 5]`), and there'd be some performance loss from using `position` as a key rather than `index`, but on the other hand some navigations will be between proposals for the same application form, and in that case performance should be ideal - matching rows will match not only their `position` but also their row header.

File: src/components/ProposalDetailPanel.tsx
Comment (jasonaowen) [reply src/components/ProposalDetailPanel.tsx:41]: I don't think this looks great:

Screen Shot 2023-03-31 at 18 44 27

It's not just the alignment, although that certainly doesn't help. I think I'll omit the icon entirely, unless someone speaks up.

File: src/components/ProposalDetailPanel.tsx
Comment (jasonaowen) [reply src/components/ProposalDetailPanel.tsx]: Here's what I have right now:

Screen Shot 2023-03-31 at 18 54 49

**Comment by jasonaowen** [comment]
Alright, I added several fixup commits to this. I'll squash them away on Monday, but for now hopefully they make seeing the changes I made in response to review comments easier.

**Comment by jasonaowen** [comment]
> I'm not currently able to run it locally because I just realized I don't know how to authenticate into the PDC from localhost (or if that's even possible)

It is possible! You should be able to plug the relevant values into your `.env.local` file; that's what I've been doing! Let's pair next week on getting you set up, @reefdog.

**Comment by jasonaowen** [comment]
> If you resolve them to your satisfaction, you can feel free to merge, or I can take a second look on Monday.

I ended up making some design changes in response to @jim-mcgowan's feedback, so yeah I'd love a second look from you please, @reefdog!

File: src/pages/ProposalDetailPage.tsx
Comment (jasonaowen) [inline src/pages/ProposalDetailPage.tsx:24]: These should live in the API helpers file.

File: src/pages/ProposalDetailPage.tsx
Comment (jasonaowen) [inline src/pages/ProposalDetailPage.tsx:72]: Extract helpers, and put them in the API file.

File: src/use-pdc-api.ts
Comment (jasonaowen) [inline src/use-pdc-api.ts:1]: Remove `use` from the file name.

File: src/pages/ProposalDetailPage.tsx
Comment (jasonaowen) [inline src/pages/ProposalDetailPage.tsx:1]: Remove `Page` from the file name - it's already in the path.

**Review by jasonaowen — COMMENTED** [review]
Before merging, I want to clean up a few things based on what I learned working on #49!

File: src/components/ProposalTable.tsx
Comment (reefdog) [reply src/components/ProposalTable.tsx:38]: Really thorough and instructive explanation, thanks @jasonaowen!

File: src/pages/ProposalDetailPage.tsx
Comment (reefdog) [inline src/pages/ProposalDetailPage.tsx:81]: When you navigate away from a proposal detail page, the title remains as set here.

Ideally, all our pages would define their titles, so this would be immediately overwritten. But we might also consider returning a cleanup function from this hook which resets the title to some default (e.g., "Philanthropy Data Commons") in case the destination page _doesn't_ have a title defined.

The downside to this is (1) the potential for the title to change twice on every navigation away slowly enough for people to notice and (2) the added boilerplate for every page that uses a title. I suspect, however, we'll want to abstract out some of this title-setting anyway.

For here and now, maybe just do the cleanup method? I think that would just mean adding `return () => document.title = 'Philanthropy Data Commons';` to the end of the hook.

File: src/use-pdc-api.ts
Comment (reefdog) [reply src/use-pdc-api.ts:31]: Ah, excellent point about `path`, and I agree it should be passed in.

`setResponse` I'm pretty sure doesn't. I read the [React guidance here](https://legacy.reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) which helped me quite a bit, but it was really the [next section on performance](https://legacy.reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) that demonstrated specifically that the state-setting function doesn't need to be passed in.

As for `fetch`, I assumed that `useOidcFetch()` was simply returning a custom fetch method that decorated the request with the token, and any determination about token freshness occurred when `fetch()` itself was called. But I'm at the edges of my understanding here. If you think the definition of `fetch` itself changes from render to render, then yes, let's keep it in the list (after that upstream bug is solved)! I still don't fully understand, but trust your knowledge of it more than mine.

File: src/components/ProposalDetailPanel.tsx
Comment (reefdog) [reply src/components/ProposalDetailPanel.tsx]: I agree with both of you, in fact! I too wanted the three-person icon, but ended up not liking it because it looks vaguely like a paw print more than a crowd — at least, I couldn't unthink it after thinking it.

I don't think the big title needs an icon, so I also agree with where you landed. But, as noted in my other comment, I want to take a crack at this header altogether after understanding Jim's feedback better.

I did notice that you used the building library icon for the EIN. I feel like without the organization label also in the same tag, we can't use such an "organization as a building" metaphor. I'm experimenting with just a small badge reading "EIN", which no doubt we'll change once we replace EIN as the most visible organization ID:

Organization header

File: src/components/ProposalDetailPanel.tsx
Comment (reefdog) [inline src/components/ProposalDetailPanel.tsx:51]: I'm pretty uncomfortable with how this incarnation makes the page feel like an Organization page, not a Proposal page. Which I understand is part of the intent, but I think it risks muddying the user's understanding of what entity they're actually looking at here.

For the demo purposes, let's just stick with it, but I think I'm going to take another pass at understanding the feedback and suggesting some more alternatives that don't relegate the proposal title itself to the panel tag area.

**Review by reefdog — CHANGES_REQUESTED** [review]
@jasonaowen This is great and could merge as-is (pending your own noted TODOs).

I did make a few changes myself (one in particular, the icon to badge change for the EIN, you can feel free to discard if you dislike it), and want to take _more_ of a crack at resolving Jim's concerns with the title/tags because I think relegating the proposal title to merely a tag with the org as the headline may confuse users in the long term. Fine for the demo, though, and heck maybe the feedback will be instructive.

A few other notes/requests but all disposable at your discretion. Sorry for the review delay!

File: src/components/ProposalDetailPanel.tsx
Comment (reefdog) [reply src/components/ProposalDetailPanel.tsx:51]: (This is more of a musing than anything else, feel free to mark as resolved.)

File: src/use-pdc-api.ts
Comment (jasonaowen) [reply src/use-pdc-api.ts:31]: > `setResponse` I'm pretty sure doesn't. I read the [React guidance here](https://legacy.reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) which helped me quite a bit, but it was really the [next section on performance](https://legacy.reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) that demonstrated specifically that the state-setting function doesn't need to be passed in, unless I'm misunderstanding our use of it.

Thank you, I hadn't read closely enough! In particular, the [old docs specifically call out `setState` as not needing to be a dependency](https://legacy.reactjs.org/docs/hooks-reference.html):

> React guarantees that `setState` function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the `useEffect` or `useCallback` dependency list.

I will fix!

> As for `fetch`, I assumed that `useOidcFetch()` was simply returning a custom fetch method that decorated the request with the token, and any determination about token freshness occurred when `fetch()` itself was called. But I'm at the edges of my understanding here. If you think the definition of `fetch` itself changes from render to render, then yes, let's keep it in the list (after that upstream bug is solved)! I still don't fully understand, but trust your knowledge of it more than mine.

Yeah, after thinking further, you're right. `useOidcFetch()` does not depend on any of the state or props we care about; it has its own internal state, but we don't need or want to re-fetch data based on any of its state - updating a token should not cause any API requests! I'll update the comment.

File: src/components/ProposalDetailPanel.tsx
Comment (jasonaowen) [reply src/components/ProposalDetailPanel.tsx]: Oh, that looks very good!

File: src/pages/ProposalDetailPage.tsx
Comment (jasonaowen) [reply src/pages/ProposalDetailPage.tsx:81]: This is a fantastic note - thank you for catching it!

I will do.

File: src/components/ProposalDetailPanel.tsx
Comment (jasonaowen) [reply src/components/ProposalDetailPanel.tsx:51]: Ah, thank you for that thoughtfulness - I implemented the feedback I got without considering that! Would you like me to make an issue for following up on that?

Comment (reefdog) [reply src/components/ProposalDetailPanel.tsx:51]: Jim and I had a meeting about the header today, and I'm trying out a few design ideas in Figma. I'll create an issue for it; what you have here can roll in without further edits.

**Review by reefdog — APPROVED** [review]
After a discussion w/@jasonaowen I'm changing my review to an "approve" to unblock his ability to unilaterally merge it in once he's made the edits he's already got queued up and/or in mind. Really happy with where this landed and agree with the edits he has planned! Thanks for all the positive iterating on it.

File: src/components/ProposalDetailPanel.tsx
Comment (reefdog) [reply src/components/ProposalDetailPanel.tsx:51]: Oh, hah, I already did make an issue for it! https://github.com/PhilanthropyDataCommons/data-viewer/issues/54

**Comment by jasonaowen** [comment]
Thanks, @reefdog!

I believe I've addressed all the feedback in this PR, and I've squashed down all the fixup commits. The tests pass, the linter is happy, the site builds, and the deploy preview looks good.

Merging!

**Pull request merged**
